File: python/packages/spore_tools/src/spore_tools/tools/params.py
"""Per-tool pydantic parameter models and a parsing helper.

Mirrors ``rust/crates/spore-core/src/tools/params.rs`` — every standard tool
deserializes its ``ToolCall.input`` dict into one of these models. Validation
failures are mapped to :class:`InvalidParameters`.
"""

from __future__ import annotations

from enum import Enum
from typing import Any, TypeVar

from pydantic import BaseModel, ConfigDict, Field, ValidationError

from spore_core.hooks import PlanArtifact
from spore_core.model import ToolCall
from spore_core.tasklist import TaskStatus

from .error import InvalidParameters


class _Params(BaseModel):
    model_config = ConfigDict(extra="forbid")


T = TypeVar("T", bound=_Params)


def parse_params(model: type[T], call: ToolCall) -> T:
    """Parse ``call.input`` into ``model`` or raise :class:`InvalidParameters`."""

    try:
        return model.model_validate(call.input)
    except ValidationError as e:
        raise InvalidParameters(reason=str(e)) from e


# ---------- Filesystem ----------


class ReadFileParams(_Params):
    path: str
    offset: int = 1
    length: int = 0
    line_numbers: bool = False


class WriteFileParams(_Params):
    path: str
    content: str
    append: bool = False


class ListDirParams(_Params):
    path: str
    recursive: bool = False
    include_ignored: bool = False


class DeleteFileParams(_Params):
    path: str


class MoveFileParams(_Params):
    src: str
    dst: str


# ---------- Exec ----------


class ExecParams(_Params):
    """Parameters for the shell-free :class:`spore_tools.tools.exec.ExecTool`:
    a program name plus a verbatim argument vector. No shell is involved."""

    command: str
    args: list[str] = Field(default_factory=list)
    timeout: int | None = None


class ShellCommandParams(_Params):
    """Parameters for the real :class:`spore_tools.tools.exec.BashCommandTool`:
    a single shell ``script`` run via ``/bin/sh -c``, with an optional working
    directory."""

    script: str
    working_dir: str | None = None
    timeout: int | None = None


class RunTestsParams(_Params):
    command: str
    working_dir: str
    timeout: int | None = None


# ---------- Search ----------


class GrepFilesParams(_Params):
    pattern: str
    path: str
    recursive: bool = False


class GrepOutputMode(str, Enum):
    """Output mode for the net-new :class:`GrepTool` (#81)."""

    CONTENT = "content"
    FILES_WITH_MATCHES = "files_with_matches"
    COUNT = "count"


class GrepParams(_Params):
    """Parameters for the net-new :class:`GrepTool` (#81). ``output_mode``
    defaults to ``content`` (``path:line:text`` per match). ``context_lines``
    adds standard ``grep -C N`` context when > 0 (#133)."""

    pattern: str
    path: str
    recursive: bool = False
    output_mode: GrepOutputMode = GrepOutputMode.CONTENT
    context_lines: int = 0  # lines of context before/after each match (default 0)


class FindFilesParams(_Params):
    glob: str
    path: str


# ---------- EditFile (#81) ----------


class EditFileParams(_Params):
    path: str
    old_string: str
    new_string: str


# ---------- SendMessage (#81) ----------


class SendMessageParams(_Params):
    content: str


# ---------- Web (#81) ----------


class WebFetchParams(_Params):
    url: str


class WebSearchParams(_Params):
    query: str


# ---------- TodoWrite (#81) ----------


class TodoStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"


class TodoItem(_Params):
    content: str
    status: TodoStatus


class TodoWriteParams(_Params):
    todos: list[TodoItem]


# ---------- Tier-3 control tools (#81) ----------


class EnterPlanModeParams(_Params):
    """``enter_plan_mode`` — context is optional (defaults to empty)."""

    context: str = ""


class ExitPlanModeParams(_Params):
    """``exit_plan_mode`` — the agent-constructed plan deserializes DIRECTLY
    into the existing :class:`~spore_core.hooks.PlanArtifact` (issue #81, Q4a —
    no stub). ``PlanArtifact.rationale`` defaults to ``""``."""

    plan: PlanArtifact


class AskUserQuestionParams(_Params):
    question: str
    options: list[str] | None = None


class AbortParams(_Params):
    reason: str


# ---------- Git ----------


class GitResetMode(str, Enum):
    HARD = "hard"
    SOFT = "soft"
    MIXED = "mixed"


class GitLogParams(_Params):
    n: int = 20
    format: str = "oneline"


class GitDiffParams(_Params):
    from_: str | None = Field(default=None, alias="from")
    to: str | None = None

    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class GitCommitParams(_Params):
    message: str
    files: list[str] = Field(default_factory=list)


class GitStatusParams(_Params):
    pass


class GitResetParams(_Params):
    target: str
    mode: GitResetMode


# ---------- HTTP ----------


class HttpGetParams(_Params):
    url: str
    headers: dict[str, Any] | None = None


class HttpPostParams(_Params):
    url: str
    body: Any
    headers: dict[str, Any] | None = None


# ---------- Subagent ----------


class SubagentParams(_Params):
    instruction: str


# ---------- TaskList (#71) ----------


class AddTaskParams(_Params):
    """``action == "add_task"`` — append a new pending task.

    ``blockers`` (#118) are ids of tasks that must be ``completed`` before this
    one runs. Optional; defaults to empty. Validated by
    :meth:`~spore_core.tasklist.TaskList.add`.
    """

    description: str
    blockers: list[int] = Field(default_factory=list)


class UpdateTaskParams(_Params):
    """``action == "update_task"`` — change status and/or description by id."""

    id: int
    status: TaskStatus | None = None
    description: str | None = None


class CompleteTaskParams(_Params):
    """``action == "complete_task"`` — mark a task completed by id."""

    id: int


class ListTasksParams(_Params):
    """``action == "list_tasks"`` — return the current list (no fields)."""


#: Maps the ``action`` discriminator to its per-action parameter model. Mirrors
#: the Rust internally-tagged ``TaskListParams`` enum.
TASK_LIST_PARAMS_BY_ACTION: dict[str, type[_Params]] = {
    "add_task": AddTaskParams,
    "update_task": UpdateTaskParams,
    "complete_task": CompleteTaskParams,
    "list_tasks": ListTasksParams,
}


# ---------- Memory (#82) ----------


class MemoryWriteParams(_Params):
    """``operation == "write"`` — append one entry to ``scope``.

    ``scope`` is a free ``str`` here (not the :class:`StorageScope` enum) so that
    a ``"local"`` scope deserializes successfully and reaches the tool body,
    where it is rejected at runtime with a recoverable error (the advertised
    schema enum omits ``local``). ``metadata`` is optional and defaults to ``{}``
    (decision C); the :class:`~spore_core.storage.MemoryEntry` type is unchanged.
    """

    scope: str
    role: str
    content: str
    metadata: dict[str, Any] = Field(default_factory=dict)


class MemoryReadParams(_Params):
    """``operation == "read"`` — read from ``scope`` (or the merged view).

    ``merged`` defaults to ``False``; ``limit`` defaults to ``50`` (decision B).
    ``scope`` is a free ``str`` for the same runtime-rejection reason as
    :class:`MemoryWriteParams`."""

    scope: str
    merged: bool = False
    limit: int = 50


#: Maps the ``operation`` discriminator to its per-operation parameter model.
#: Mirrors the Rust internally-tagged ``MemoryToolParams`` enum.
MEMORY_PARAMS_BY_OPERATION: dict[str, type[_Params]] = {
    "write": MemoryWriteParams,
    "read": MemoryReadParams,
}


def parse_memory_params(call: ToolCall) -> tuple[str, _Params]:
    """Parse the ``operation`` discriminator and dispatch to the matching model.

    Returns ``(operation, params)``. Raises :class:`InvalidParameters` if the
    input is not an object, ``operation`` is missing/unknown, or the
    per-operation fields fail validation. The ``operation`` field is stripped
    before per-operation validation so the ``extra="forbid"`` models do not
    reject it.
    """
    input_ = call.input
    if not isinstance(input_, dict):
        raise InvalidParameters(reason="input must be a JSON object")
    operation = input_.get("operation")
    if not isinstance(operation, str):
        raise InvalidParameters(reason="missing or non-string `operation`")
    model = MEMORY_PARAMS_BY_OPERATION.get(operation)
    if model is None:
        raise InvalidParameters(reason=f"unknown operation `{operation}`")
    fields = {k: v for k, v in input_.items() if k != "operation"}
    try:
        params = model.model_validate(fields)
    except ValidationError as e:
        raise InvalidParameters(reason=str(e)) from e
    return operation, params


def parse_task_list_params(call: ToolCall) -> tuple[str, _Params]:
    """Parse the ``action`` discriminator and dispatch to the matching model.

    Returns ``(action, params)``. Raises :class:`InvalidParameters` if the input
    is not an object, ``action`` is missing/unknown, or the per-action fields
    fail validation. The ``action`` field is stripped before per-action
    validation so the ``extra="forbid"`` models do not reject it.
    """
    input_ = call.input
    if not isinstance(input_, dict):
        raise InvalidParameters(reason="input must be a JSON object")
    action = input_.get("action")
    if not isinstance(action, str):
        raise InvalidParameters(reason="missing or non-string `action`")
    model = TASK_LIST_PARAMS_BY_ACTION.get(action)
    if model is None:
        raise InvalidParameters(reason=f"unknown action `{action}`")
    fields = {k: v for k, v in input_.items() if k != "action"}
    try:
        params = model.model_validate(fields)
    except ValidationError as e:
        raise InvalidParameters(reason=str(e)) from e
    return action, params


__all__ = [
    "AddTaskParams",
    "CompleteTaskParams",
    "ListTasksParams",
    "MEMORY_PARAMS_BY_OPERATION",
    "MemoryReadParams",
    "MemoryWriteParams",
    "TASK_LIST_PARAMS_BY_ACTION",
    "UpdateTaskParams",
    "parse_memory_params",
    "parse_task_list_params",
    "AbortParams",
    "AskUserQuestionParams",
    "DeleteFileParams",
    "EditFileParams",
    "EnterPlanModeParams",
    "ExitPlanModeParams",
    "FindFilesParams",
    "GitCommitParams",
    "GitDiffParams",
    "GitLogParams",
    "GitResetMode",
    "GitResetParams",
    "GitStatusParams",
    "ExecParams",
    "GrepFilesParams",
    "GrepOutputMode",
    "GrepParams",
    "HttpGetParams",
    "HttpPostParams",
    "ListDirParams",
    "MoveFileParams",
    "ReadFileParams",
    "RunTestsParams",
    "SendMessageParams",
    "ShellCommandParams",
    "SubagentParams",
    "TodoItem",
    "TodoStatus",
    "TodoWriteParams",
    "WebFetchParams",
    "WebSearchParams",
    "WriteFileParams",
    "parse_params",
]

File: go/spore-core/tools/params.go
// Per-tool input parameter structs (issue #5).
//
// One struct per tool. Each maps 1:1 onto the Rust serde struct in
// rust/crates/spore-core/src/tools/params.rs. Wire-compatible JSON tags use
// snake_case.

package tools

import (
	"encoding/json"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/storage"
)

// StorageScope is re-exported from the storage package (its canonical home is
// promptassembly; storage re-exports it). The MemoryTool params name it without
// forcing every params caller to import storage directly.
type StorageScope = storage.StorageScope

// Scope constants re-exported for use at the MemoryTool call site.
const (
	StorageScopeUser    = storage.StorageScopeUser
	StorageScopeProject = storage.StorageScopeProject
	StorageScopeLocal   = storage.StorageScopeLocal
)

// ----- Filesystem -----

// ReadFileParams are the parameters for ReadFileTool. With all optional fields
// at their defaults (Offset=nil, Length=0, LineNumbers=false) the read is
// byte-identical to reading the whole file — no header, no line numbers (#132).
// Any non-default param prepends a one-line header
// "[lines {start}–{end} of {total}]" (U+2013 en-dash).
//
// Offset is a *uint64 (pointer) so we can distinguish "not provided" (nil,
// default=1) from explicitly-zero (0, which is a recoverable error per spec).
type ReadFileParams struct {
	Path string `json:"path"`
	// 1-indexed start line. nil means "not provided" (default = 1).
	// Explicitly 0 is a recoverable error.
	Offset *uint64 `json:"offset,omitempty"`
	// Max lines to return. 0 = no limit / read to EOF (default 0).
	// A length that runs past EOF silently returns through the last line.
	Length uint64 `json:"length,omitempty"`
	// When true, prefix each returned line with its 1-indexed number,
	// right-padded to the digit-width of the file's total line count.
	LineNumbers bool `json:"line_numbers,omitempty"`
}

type WriteFileParams struct {
	Path    string `json:"path"`
	Content string `json:"content"`
	Append  bool   `json:"append,omitempty"`
}

type ListDirParams struct {
	Path           string `json:"path"`
	Recursive      bool   `json:"recursive,omitempty"`
	IncludeIgnored bool   `json:"include_ignored,omitempty"` // default false
}

type DeleteFileParams struct {
	Path string `json:"path"`
}

type MoveFileParams struct {
	Src string `json:"src"`
	Dst string `json:"dst"`
}

// ----- Exec -----

// ExecParams are the parameters for the shell-free ExecTool: a program name
// plus a verbatim argument vector. No shell is involved, so the args are passed
// literally (no pipes, redirects, globbing, or $(...)).
type ExecParams struct {
	Command string   `json:"command"`
	Args    []string `json:"args,omitempty"`
	// Timeout in whole seconds. *uint64 so the absence of the field is
	// distinguishable from 0.
	Timeout *uint64 `json:"timeout,omitempty"`
}

// ShellCommandParams are the parameters for the real BashCommandTool: a single
// shell script run via /bin/sh -c, with an optional working directory.
type ShellCommandParams struct {
	Script string `json:"script"`
	// Optional working directory; the only path that gets sandbox validation.
	WorkingDir string `json:"working_dir,omitempty"`
	// Timeout in whole seconds. *uint64 so the absence of the field is
	// distinguishable from 0.
	Timeout *uint64 `json:"timeout,omitempty"`
}

type RunTestsParams struct {
	Command    string  `json:"command"`
	WorkingDir string  `json:"working_dir"`
	Timeout    *uint64 `json:"timeout,omitempty"`
}

// ----- Search -----

type GrepFilesParams struct {
	Pattern   string `json:"pattern"`
	Path      string `json:"path"`
	Recursive bool   `json:"recursive,omitempty"`
}

type FindFilesParams struct {
	Glob string `json:"glob"`
	Path string `json:"path"`
}

// ----- EditFile (#81, new) -----

// EditFileParams are the parameters for EditFileTool: replace the FIRST and
// ONLY occurrence of OldString with NewString in the file at Path. The match
// must be unique — an absent or non-unique OldString is a recoverable error.
type EditFileParams struct {
	Path      string `json:"path"`
	OldString string `json:"old_string"`
	NewString string `json:"new_string"`
}

// ----- Grep (#81, new — output modes) -----

// GrepOutputMode selects GrepTool's output shape. One of "content" (default),
// "files_with_matches", or "count".
type GrepOutputMode string

const (
	// GrepOutputContent emits each matching line as `path:line:text` (default).
	GrepOutputContent GrepOutputMode = "content"
	// GrepOutputFilesWithMatches emits the distinct file paths with a match.
	GrepOutputFilesWithMatches GrepOutputMode = "files_with_matches"
	// GrepOutputCount emits `path:count` for each file with matches.
	GrepOutputCount GrepOutputMode = "count"
)

// GrepParams are the parameters for the net-new GrepTool. Distinct from
// GrepFilesParams: adds OutputMode (defaulting to content) and ContextLines.
type GrepParams struct {
	Pattern      string         `json:"pattern"`
	Path         string         `json:"path"`
	Recursive    bool           `json:"recursive,omitempty"`
	OutputMode   GrepOutputMode `json:"output_mode,omitempty"`
	ContextLines uint32         `json:"context_lines,omitempty"` // default 0
}

// UnmarshalJSON applies the default OutputMode=content when absent or empty.
func (p *GrepParams) UnmarshalJSON(data []byte) error {
	type alias GrepParams
	a := alias{}
	if err := json.Unmarshal(data, &a); err != nil {
		return err
	}
	if a.OutputMode == "" {
		a.OutputMode = GrepOutputContent
	}
	*p = GrepParams(a)
	return nil
}

// ----- SendMessage (#81, new) -----

// SendMessageParams are the parameters for SendMessageTool.
type SendMessageParams struct {
	Content string `json:"content"`
}

// ----- TodoWrite (#81, new) -----

// TodoStatus is one of "pending" | "in_progress" | "completed".
type TodoStatus string

const (
	TodoStatusPending    TodoStatus = "pending"
	TodoStatusInProgress TodoStatus = "in_progress"
	TodoStatusCompleted  TodoStatus = "completed"
)

// TodoItem is a single todo entry managed by TodoWriteTool.
type TodoItem struct {
	Content string     `json:"content"`
	Status  TodoStatus `json:"status"`
}

// TodoWriteParams are the parameters for TodoWriteTool: the agent supplies the
// FULL desired todo list, which replaces the persisted list wholesale.
type TodoWriteParams struct {
	Todos []TodoItem `json:"todos"`
}

// ----- WebFetch / WebSearch (#81, new) -----

// WebFetchParams are the parameters for WebFetchTool.
type WebFetchParams struct {
	URL string `json:"url"`
}

// WebSearchParams are the parameters for WebSearchTool.
type WebSearchParams struct {
	Query string `json:"query"`
}

// ----- Tier 3: plan / clarify / abort (#81, new) -----

// EnterPlanModeParams are the parameters for EnterPlanModeTool. Context is
// optional (defaults to empty).
type EnterPlanModeParams struct {
	Context string `json:"context,omitempty"`
}

// ExitPlanModeParams are the parameters for ExitPlanModeTool. The agent supplies
// the plan as a structured object that deserializes DIRECTLY into the existing
// sporecore.PlanArtifact (issue #81, Q4a — no stub).
type ExitPlanModeParams struct {
	Plan sporecore.PlanArtifact `json:"plan"`
}

// AskUserQuestionParams are the parameters for AskUserQuestionTool. Options is
// nil for a free-form clarification.
type AskUserQuestionParams struct {
	Question string    `json:"question"`
	Options  *[]string `json:"options,omitempty"`
}

// AbortParams are the parameters for AbortTool.
type AbortParams struct {
	Reason string `json:"reason"`
}

// ----- Git -----

type GitLogParams struct {
	N      uint32 `json:"n"`
	Format string `json:"format"`
}

// UnmarshalJSON applies the defaults n=20, format="oneline".
func (p *GitLogParams) UnmarshalJSON(data []byte) error {
	type alias GitLogParams
	a := alias{N: 20, Format: "oneline"}
	if err := json.Unmarshal(data, &a); err != nil {
		return err
	}
	if a.N == 0 {
		a.N = 20
	}
	if a.Format == "" {
		a.Format = "oneline"
	}
	*p = GitLogParams(a)
	return nil
}

type GitDiffParams struct {
	From *string `json:"from,omitempty"`
	To   *string `json:"to,omitempty"`
}

type GitCommitParams struct {
	Message string   `json:"message"`
	Files   []string `json:"files,omitempty"`
}

type GitStatusParams struct{}

// GitResetMode is one of "hard" | "soft" | "mixed".
type GitResetMode string

const (
	GitResetHard  GitResetMode = "hard"
	GitResetSoft  GitResetMode = "soft"
	GitResetMixed GitResetMode = "mixed"
)

type GitResetParams struct {
	Target string       `json:"target"`
	Mode   GitResetMode `json:"mode"`
}

// ----- HTTP -----

type HttpGetParams struct {
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
}

type HttpPostParams struct {
	URL     string            `json:"url"`
	Body    json.RawMessage   `json:"body"`
	Headers map[string]string `json:"headers,omitempty"`
}

// ----- Subagent -----

type SubagentParams struct {
	Instruction string `json:"instruction"`
}

// ----- Memory (#82) -----

// MemoryOperation is the operation discriminator for MemoryToolParams. One of
// "write" | "read".
type MemoryOperation string

const (
	// MemoryOperationWrite appends one MemoryEntry to a scope.
	MemoryOperationWrite MemoryOperation = "write"
	// MemoryOperationRead returns the most-recent entries for a scope (or the
	// cross-scope merged view).
	MemoryOperationRead MemoryOperation = "read"
)

// MemoryDefaultReadLimit is the recency cap applied to a read when `limit` is
// omitted (decision B).
const MemoryDefaultReadLimit = 50

// MemoryToolParams are the parameters for the MemoryTool. It mirrors the Rust
// internally-tagged MemoryToolParams enum (tagged on `operation`): an
// `operation` discriminator plus the union of per-operation fields. `scope` is
// explicit and required on BOTH operations.
//
// Per-operation fields (validated in the tool, not by serde):
//   - write: scope (required), role (required), content (required),
//     metadata (optional, defaults to {})
//   - read:  scope (required), merged (optional, defaults false),
//     limit (optional, defaults to MemoryDefaultReadLimit)
//
// Scope is decoded as a raw promptassembly StorageScope string ("user" /
// "project" / "local"). StorageScopeLocal decodes fine here so a bad-scope call
// reaches the tool body, where it is rejected at runtime with a recoverable
// error (the advertised schema enum omits "local").
//
// Limit uses a *int so an omitted `limit` (nil) is distinguishable from an
// explicit 0; the tool applies MemoryDefaultReadLimit when nil. Metadata uses
// json.RawMessage; UnmarshalJSON fills it with {} when absent or null so the
// stored entry's metadata default matches Rust's serde default.
type MemoryToolParams struct {
	Operation MemoryOperation `json:"operation"`
	Scope     StorageScope    `json:"scope"`
	Role      string          `json:"role,omitempty"`
	Content   string          `json:"content,omitempty"`
	Metadata  json.RawMessage `json:"metadata,omitempty"`
	Merged    bool            `json:"merged,omitempty"`
	Limit     *int            `json:"limit,omitempty"`
}

// UnmarshalJSON fills Metadata with the empty object {} when the field is absent
// or null, mirroring serde's #[serde(default)] on the write metadata param.
func (p *MemoryToolParams) UnmarshalJSON(data []byte) error {
	type alias MemoryToolParams
	var a alias
	if err := json.Unmarshal(data, &a); err != nil {
		return err
	}
	if len(a.Metadata) == 0 || string(a.Metadata) == "null" {
		a.Metadata = json.RawMessage("{}")
	}
	*p = MemoryToolParams(a)
	return nil
}

// ----- TaskList (#71) -----

// TaskListAction is the action discriminator for TaskListParams. One of
// "add_task" | "update_task" | "complete_task" | "list_tasks".
type TaskListAction string

const (
	TaskListActionAddTask      TaskListAction = "add_task"
	TaskListActionUpdateTask   TaskListAction = "update_task"
	TaskListActionCompleteTask TaskListAction = "complete_task"
	TaskListActionListTasks    TaskListAction = "list_tasks"
)

// TaskListParams are the parameters for the TaskListTool. Mirrors the Rust
// internally-tagged TaskListParams enum: an `action` discriminator plus the
// union of per-action fields. Pointer fields distinguish present-vs-absent so
// update_task can apply status and/or description independently.
//
// Field requirements per action (validated in the tool, not by serde):
//   - add_task:      description (required); blockers (optional, defaults empty)
//   - update_task:   id (required); status and/or description (both optional)
//   - complete_task: id (required)
//   - list_tasks:    no fields
//
// Blockers (#118) are the ids that must be completed before the added task runs;
// omitting the field defaults to empty. They are threaded through to
// TaskList.Add, which validates them.
type TaskListParams struct {
	Action      TaskListAction        `json:"action"`
	Description *string               `json:"description,omitempty"`
	ID          *uint32               `json:"id,omitempty"`
	Status      *sporecore.TaskStatus `json:"status,omitempty"`
	Blockers    []uint32              `json:"blockers,omitempty"`
}
